module-4/backend/week_three.md

```diff
@@ -5,15 +5,119 @@ Fork this respository. Answer the questions to the best of your ability. Try to
+  - In `server.js`
+```javascript
+const express = require('express')
+const app = express()
+const bodyParser = require('body-parser')
+
+app.use(bodyParser.json())
+app.use(bodyParser.urlencoded({ extended: true }))
+
+app.set('port', process.env.PORT || 3000)
+
+app.get('/api-url', function(req, res) {
+  // do a thing to get all models
+  // display as json
+})
+
+app.get('/api-url/:id', function(req, res) {
+  // do a thing to get a specific model
+  // display as json
+})
+app.post('/api-url', function(req, res) {
+  // do a thing to create a new model
+  // display as json
+})
+```
+  - or refactored...
+  - in `server.js`
+```javascript
+const express = require('express')
+const app = express()
+const bodyParser = require('body-parser')
+const ExamplesController = require(./lib/controllers/examples) 
+
+app.use(bodyParser.json())
+app.use(bodyParser.urlencoded({ extended: true }))
+
+app.set('port', process.env.PORT || 3000)
+
+app.get('/api-url', ExamplesController.index)
+
+app.get('/api-url/:id', ExamplesController.show)
+
+app.post('/api-url', ExamplesController.create)
+```
+  - And in `lib/controllers/example.js`
+```javascript
+const Example = require('../models/example') // For SQL queries
+
+const index = (req, res) => {
+  // do a thing to get a all models
+  // display as json
+}
+
+const show = (req, res) => {
+  // do a thing to get a specific model
+  // display as json
+}
+
+const create = (req, res) => {
+  // do a thing to create a new model
+  // display as json
+}
+
+module.exports = {
+  index, 
+  show,
+  create
+}
+```
 4. What do we use Knex for?
+  - We use Knex to make queries to our DB and and return objects that can be accessed/interacted with elsewhere in the code.
+  - It's our ORM
 5. How could you organize your code to follow the MVC design pattern in your Quantified Self project?
+  - See #3
 6. How do you execute raw SQL in node?
+```javascript
+const environment = process.env.NODE_ENV || 'development'
+const configuration = require('./knexfile')[environment]
+const database = require('knex')(configuration)
+
+return database.raw('SELECT * FROM examples;')
+.then((data) => {
+  // Do a thing with the returned promise. 
+  // data.rows = array of database model objects
+})
+```
 7. What are some advantages to having your front end and back end code live in separate applications? What are some disadvantages?
+  - Advantages: 
+    - Connecting by API allows for a large variety of languages to be used for both frontend and backend languages easily
+    - More abrstraction by responsibility, frontend vs backend
+  
+  - Disadvantages: 
+    - Host two seperate apps on two servers
 
 #### Review  
 
 8. Describe DNS.
+  - Domain Name Servers. This allows for websites to be looked up by the IP address associated with that url
 9. What does writing clean code mean to you?
+  - Code that is flexible and can be reused.
```

Comment:
good list here - i encourage you to be even more specific here and even potentially give examples, what does a "hard-coded" method look like? why don't we want those methods?